xplique/attributions/smoothgrad.py

```diff
@@ -73,45 +73,72 @@ def explain(self,
+            # group by inputs and compute the average gradient
+            gradients = tf.reshape(gradients, (-1, self.nb_samples, *gradients.shape[1:]))
+            reduced_gradients = self._reduce_gradients(gradients)
 
-                # apply noise
-                batch_noisy_inputs = single_input + batch_noises
-                repeated_targets = repeat_labels(single_target[tf.newaxis, :], nb_noises)
+            smoothed_gradients = reduced_gradients if smoothed_gradients is None else tf.concat(
+                [smoothed_gradients, reduced_gradients], axis=0)
 
-                # compute the gradient of each noisy samples generated
-                batch_gradients = gradient(self.model, batch_noisy_inputs, repeated_targets)
+        return smoothed_gradients
+
+    @staticmethod
+    def _get_noisy_mask(shape: Tuple[int, int, int, int],
+                        noise: float) -> tf.Tensor:
+        """
+        Create a random noise mask of the specified shape.
+
+        Parameters
+        ----------
+        shape
+            Desired shape, dimension of one sample.
+        noise
+            Scalar, noise used as standard deviation of a normal law centered on zero.
 
-                # mean over a batch of gradients for one input
-                reduced_gradients = self._reduce_gradients(batch_gradients)
+        Returns
+        -------
+        noisy_mask
+            Noise mask of the specified shape.
+        """
+        return tf.random.normal(shape, 0.0, noise, dtype=tf.float32)
 
-                # accumulate weighted reduced gradient for weighted mean
-                accumulated_reduced_gradients += nb_noises * reduced_gradients
+    @staticmethod
+    @tf.function
+    def _apply_noise(inputs: tf.Tensor,
+                     noisy_mask: tf.Tensor) -> tf.Tensor:
+        """
+        Duplicate the samples and apply a noisy mask to each of them.
 
-            # weighted mean of the mean of batch of gradients for one input
-            reduced_reduced_gradients = accumulated_reduced_gradients / self.nb_samples
+        Parameters
+        ----------
+        inputs
+            Input samples to be explained.
+        noisy_mask
+            Mask of random noise to apply on a set of interpolations points. With S the number of
+            samples, W & H the sample dimensions and C the number of channels.
 
-            if smoothed_gradients is None:
-                smoothed_gradients = reduced_reduced_gradients[tf.newaxis, :]
-            else:
-                smoothed_gradients = tf.concat(
-                    [smoothed_gradients, reduced_reduced_gradients[tf.newaxis, :]], axis=0)
+        Returns
+        -------
+        noisy_inputs
+            Duplicated inputs with noisy mask applied.
+        """
+        nb_samples = len(noisy_mask)
 
-        return smoothed_gradients
+        noisy_inputs = tf.repeat(tf.expand_dims(inputs, axis=1), repeats=nb_samples, axis=1)
```

Comment:
Do we agree that this creates a Tensor of shape `(batch_size, nb_samples, *inputs.shape[1:])`? For high resolution images, the `nb_sample` parameter should be higher than usual image, combined with the important `inputs.shape`, does it not create a really huge tensor even with a `batch_size` of 1?

My point here is that, for me there may be models or dataset for which the tensor created with `nb_samples` necessary for convergence cannot be stored in the memory.